<a href="https://colab.research.google.com/github/xDanilog27/Bussiness-Inteligent/blob/main/Reto_2_%C3%81rboles_de_Decisi%C3%B3n_(Aprendizaje_Supervisado).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Reto 2 - Árboles de Decisión (Aprendizaje Supervisado)**

0. Cargar Librerias

In [ ]:
# Librerias Cientificas
import pandas as pd
import numpy as np

# Librerias Arbol
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix

1. Cargar los Datos

In [ ]:
nxl="/content/2. TechSalesReport_Int_L.xlsx"
XDB=pd.read_excel(nxl,sheet_name=0) #BASE DE DATOS

# Mapear 'Hardware' a 0 y 'Software' a 1 en la columna 'Business'
XDB['Business'] = XDB['Business'].map({'Hardware': 0, 'Software': 1})

# Mapear 'Yes' a 1 y 'No' a 0 en la columna 'College'
XDB['College'] = XDB['College'].map({'Yes': 1, 'No': 0})

# Seleccionar las variables
XD=XDB[['Business','Age','Female','Years','Certificates','Feedback','Salary','NPS']]
yd=XDB[['College']]



NameError: name 'pd' is not defined

Descripción de las Variables:

De acuerdo con la base de datos de TechSalesReport_Int_L, esta posee un total de 21990 dato totales, Divididos en 4470 no poseen título universitario y 17518 poseen título universitario. Cada uno de los empleados estará descrito en función de 8 variables:
•	Years: Representa el número de años que un empleado ha trabajado en la empresa.
•	College (Variable Objetivo): Indica si el empleado posee un título universitario de cuatro años (Sí o No).
•	Certificates: Representa el número de certificaciones profesionales relevantes que ha obtenido cada empleado.
•	Feedback: Representa el puntaje de retroalimentación promedio que cada empleado recibe de sus compañeros y supervisor en la evaluación anual de 360 grados.
•	Salary: Representa el sueldo base anual de cada empleado.
•	NPS (Net Promoter Score): Es un indicador clave de la satisfacción y lealtad del cliente.
•	Business: Representa el tipo de negocio en el que trabaja el empleado, Hardware o Software. Se codificará como 0 para "Hardware" y 1 para "Software".
•	Age: Representa la edad del empleado.
•	Female: Representa el genero del empleado.


2. Implementacion del Modelo





In [ ]:
mar=DecisionTreeClassifier(criterion='gini', max_depth=4) #mar= Modelo de Arbol -- Max_Depth=maximo 4 porque mas es muy largo
mar.fit(XD,yd) # Establecer una relacion de los datos de entrada y la salida

# Se procede con la grafica
from pydotplus import graph_from_dot_data # Crea la grafica
from sklearn.tree import export_graphviz # Exporta a PNG, JPEG y TIFF

ve=['Business','Age','Female','Years','Certificates','Feedback','Salary','NPS'] #Nombre para el arbol - ve= Variables de entrada
dot_data=export_graphviz(mar, feature_names=ve) # Crear el grafico
graph=graph_from_dot_data(dot_data)
graph.write_png('arbol.png')

True

3. Pronóstico del título Universitario para las personas agrupadas en la base de datos de pronóstico



In [ ]:
XDB0 = pd.read_excel(nxl, sheet_name=1)
XDB0.head(100)

XDB0['Business'] = XDB0['Business'].map({'Hardware': 0, 'Software': 1})

# Crea la base de datos a pronosticar
XD0 = XDB0[['Business', 'Age', 'Female', 'Years', 'Certificates', 'Feedback', 'Salary', 'NPS']]
yd0 = XDB0[['College']]

# No es necesario convertir a NumPy si solo se usa para predicción
# XD0 = np.array(XD0)
# XD0 = XD0[filas]  # Eliminado porque 'filas' no está definido

# Le calculamos la preaprobación a los solicitantes
y_pred = mar.predict(XD0)
print(y_pred)

# Se crea el informe en Excel
df = pd.DataFrame(np.column_stack((XD0, y_pred)))  # No se necesita filas, ya que se usan todas las filas.
df.columns = ['Business', 'Age', 'Female', 'Years', 'Certificates', 'Feedback', 'Salary', 'NPS', 'College']
df.to_excel('Predicciones.xlsx')

[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]


In [ ]:
ydp=mar.predict(XD) #Pronostio sobre la base de datos original
cm=confusion_matrix(yd,ydp) # Matriz de confusion
print("La matriz de consion es:\n",cm)

VN=cm[0,0]
FP=cm[0,1]
FN=cm[1,0]
VP=cm[1,1]

Exactitutud=(VP+VN)/(VP+VN+FP+FN)
print("La exactitud es:",Exactitutud*100)

Sensibilidad=VP/(VP+FN)
print("La sensibilidad es:",Sensibilidad*100)

Especificidad=VN/(VN+FP)
print("La especificidad es:",Especificidad*100)

Precision=VP/(VP+FP)
print("La precision es:",Precision*100)

TasaError=1-Exactitutud
print("La tasa de error es:",TasaError*100)

PrediccionNegativa=VN/(VN+FN)
print("La prediccion negativa es:",PrediccionNegativa*100)

PrediccionPositiva=VP/(VP+FP)
print("La prediccion positiva es:",PrediccionPositiva*100)


La matriz de consion es:
 [[  114  4356]
 [   69 17451]]
La exactitud es: 79.87721691678036
La sensibilidad es: 99.60616438356165
La especificidad es: 2.5503355704697985
La precision es: 80.02476269087909
La tasa de error es: 20.12278308321964
La prediccion negativa es: 62.295081967213115
La prediccion positiva es: 80.02476269087909


**Análisis de Resultados**
La matriz de confusión revela predominancia de los valores Verdaderos Positivos (VP) (17451) y Falsos Positivos (FP) (4356), lo que indica que el modelo tiende a clasificar la mayoría de las instancias como pertenecientes a la clase positiva que son las personas con título universitario.

El modelo logró alcanzar valores superiores al 75% en promedio para los índices de exactitud (79.88%), sensibilidad (99.61%), precisión (80.02%), y especificidad (2.55%). Es importante notar que la especificidad, aunque incluida, es extremadamente baja, lo que indica que el modelo tiene dificultades para identificar correctamente a las personas sin título universitario. Esto es un reflejo directo del desbalance de clases y del sesgo del modelo hacia la predicción de la clase mayoritaria.